acvm_backend_barretenberg/src/acvm_interop/pwg.rs

```diff
@@ -162,3 +165,279 @@ impl PartialWitnessGenerator for Barretenberg {
         Ok(OpcodeResolution::Solved)
     }
 }
+
+// TODO: alter this method so that it only processes one hash per level rather than overriding
+// the one of leaves for each level of the hash path
+fn calculate_merkle_root(
+    hash_func: impl Fn(&FieldElement, &FieldElement) -> FieldElement,
+    hash_path: Vec<&FieldElement>,
+    index: &FieldElement,
+    leaf: &FieldElement,
+) -> FieldElement {
+    let mut index_bits: Vec<bool> = index.bits();
+    index_bits.reverse();
+
+    assert!(
+        hash_path.len() <= index_bits.len(),
+        "hash path exceeds max depth of tree"
+    );
+    index_bits.into_iter().zip(hash_path.into_iter()).fold(
+        *leaf,
+        |current_node, (path_bit, path_elem)| {
+            let (left, right) = if !path_bit {
+                (&current_node, path_elem)
+            } else {
+                (path_elem, &current_node)
+            };
+            hash_func(left, right)
+        },
+    )
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I don't see the need to bring these tests and merkle root calculation into `pwg.rs`. I think pwg.rs should only contain the PartialWitnessGenerator. If we want to remove the merkle.rs file from the `pwg` folder is there a way we could have a `Merkle` trait that impls on Barretenberg similarly to Pedersen and Schnorr (although this may be more confusing as we are not using barretenberg for our merkle logic, only the hash function).

Reply:
I've moved this back up into `pwg/merkle.rs` .